### PR TITLE
Have the right horicromtitude

### PR DESCRIPTION
**Pull request by mcovarr**
Confirm the distribution of workflows in a horizontal Cromwell Centaur run looks as symmetric as expected.

### CODE REVIEW

File: services/src/main/scala/cromwell/services/metadata/WorkflowProcessingEventPublishing.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+import akka.actor.ActorRef
+import common.util.VersionUtil
+import cromwell.core.WorkflowId
+import cromwell.core.WorkflowProcessingEvents.EventKey._
+import cromwell.core.WorkflowProcessingEvents._
+import cromwell.services.metadata.MetadataService.PutMetadataAction
+
+import scala.util.Random
+
+object WorkflowProcessingEventPublishing {
```
Comment (kshakir) [inline services/src/main/scala/cromwell/services/metadata/WorkflowProcessingEventPublishing.scala]: Still reviewing, but should this go in `engine`? It seems to a) only be used by engine, and b) is the **only** reason `services` now has to depend on `core`.

File: centaur/src/main/scala/centaur/CromwellTracker.scala [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+  def assertHoricromtality(): Unit = {
+    counts foreach { case (id, count) => log.info(s"horicromtal count {}: {}", id, count) }
+
+    val runs = counts.values.sum
+    val expected: Array[Double] = (1 to backendCount) map { _ => runs.toDouble / backendCount } toArray
+    val actual: Array[Long] = counts.values map { _.toLong } toArray
+
+    val observedSignificance = new ChiSquareTest().chiSquareTest(expected, actual)
+    log.info(s"configured/observed horicromtal significance levels: {}/{}", configuredSignificance, observedSignificance)
```
Comment (kshakir) [inline centaur/src/main/scala/centaur/CromwellTracker.scala]: As is the `s""` isn't needed. Alternatively, after switching to Scala Logging, consider using the [`f""` interpolator](https://docs.scala-lang.org/overviews/core/string-interpolation.html#the-f-interpolator) to shorten the decimals to something like `%.4f` or so.

The `f""` could be used in the exception message below also.

File: centaur/src/main/scala/centaur/CromwellTracker.scala [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+
+
+case class CromwellTracker(backendCount: Int, configuredSignificance: Double) {
+  val log = LoggerFactory.getLogger(getClass)
+  var counts: Map[String, Int] = Map()
+  def track(metadata: WorkflowMetadata): Unit = {
+
+    // Get all the workflow processing events for this workflow and pick out the "PickedUp" events. There can be multiple of
+    // these for a workflow if Cromwell was restarted while it was running.
+    val pickupEvents = parse(metadata.value) flatMap { _.hcursor.get[Array[WorkflowProcessingEvent]](WorkflowProcessingEvents.ProcessingEventsKey) }
```
Comment (kshakir) [inline centaur/src/main/scala/centaur/CromwellTracker.scala]: ToL-ish: I find [Circe Optics](https://circe.github.io/circe/optics.html) just-as-dense-to-read-as-HCursors while being prettier. I don't really know what the other tradeoffs would be. Here's what I think the optics version would look like.

```suggestion
    val pickupEvents = parse(metadata.value) map {
      root.workflowProcessingEvents.each.as[WorkflowProcessingEvent].getAll(_).toArray
    }
```

Note: Would also require `import io.circe.optics.JsonPath._` and later a `implicit lazy val workflowProcessingEventEncoder: Encoder[WorkflowProcessingEvent] = deriveEncoder` as optics support round-trips.

File: project/Dependencies.scala [hunk 4]
```diff
@@ -442,7 +443,8 @@ object Dependencies {
 
   val centaurDependencies = List(
     "com.github.kxbmap" %% "configs" % configsV,
-    "com.google.cloud" % "google-cloud-bigquery" % googleCloudCoreV % IntegrationTest
+    "com.google.cloud" % "google-cloud-bigquery" % googleCloudCoreV % IntegrationTest,
+    "org.apache.commons" % "commons-math3" % commonsMathV
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply project/Dependencies.scala]: `it:` dependencies?

[the remark this answers is not in the document]
Comment (kshakir) [reply project/Dependencies.scala]: `it:` == "IntegrationTest" in Maven/[SBT](https://www.scala-sbt.org/1.0/docs/Testing.html#Integration+Tests)-terms, not in Cromwell-terms.

File: src/ci/resources/centaur_application.conf [hunk 5]
```diff
@@ -10,6 +10,12 @@ centaur {
 
   standardTestCasePath: ${CROMWELL_BUILD_CENTAUR_TEST_DIRECTORY}
 
+  # horicromtal {
+  #   assert: true # default false
+  #   backend-count: 2 # no default, must be specified if assert = true
+  #   significance-level: 0.05 # default 0.05
+  #}
+
```
[the remark this answers is not in the document]
Comment (mcovarr) [reply src/ci/resources/centaur_application.conf]: Vestiges of an earlier dead-end factoring, I'll delete.

**Comment by mcovarr** [comment]
also may or may not satisfy  #4668

File: centaur/src/main/scala/centaur/CromwellTracker.scala [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    counts foreach { case (id, count) => logger.info(s"horicromtal count {}: {}", id, count) }
+
+    val runs = counts.values.sum
+    val expected: Array[Double] = (1 to backendCount) map { _ => runs.toDouble / backendCount } toArray
+    val actual: Array[Long] = counts.values map { _.toLong } toArray
+
+    val observedSignificance = new ChiSquareTest().chiSquareTest(expected, actual)
+    logger.info(f"configured/observed horicromtal significance levels: $configuredSignificance%.4f/$observedSignificance%.4f", configuredSignificance, observedSignificance)
+
+    if (observedSignificance < configuredSignificance) {
```
Comment (aednichols) [inline centaur/src/main/scala/centaur/CromwellTracker.scala]: Very cool!

How often are we likely to get random anomalies with the current significance? Not a PR-level concern, just curious how you reasoned about it.

Comment (mcovarr) [reply centaur/src/main/scala/centaur/CromwellTracker.scala]: I used the customary p < 0.05 significance for now. N is pretty larger here (> 300) and in practice the lowest significance I've seen was > 0.42. As long Cromwell really is running horizontally I don't think we should have random failures.

**Pull request merged**